### PR TITLE
gh-116987: Remove re pattern check in inspect.findsource as it's not needed anymore

### LINKED ISSUE #116987: inspect.getsource does not work for a class-defined code object
# Bug report

### Bug description:

In the following MRE, `code.co_firstlineno` correctly returns 4, the first line number of the class definition, but `getsource(code)` returns the content of the entire file:
```python
import sys
from inspect import getsource

class A:
    code = sys._getframe(0).f_code
    print(code.co_firstlineno)
    print(getsource(code))
```
This is because `inspect.findsource` uses a regex pattern that does not match `class`:
https://github.com/python/cpython/blob/a8e93d3dca086896e668b88b6c5450eaf644c0e7/Lib/inspect.py#L1155

### CPython versions tested on:

3.12

### Operating systems tested on:

Linux, Windows

### Linked PRs
* gh-117025

### PR DESCRIPTION
**Pull request by gaogaotiantian**
The pattern check code was added more than 20 years ago when `code.co_firstlineno` could be inaccurate (for multiline definitions maybe?). However, it's not needed anymore because `code.co_firstlineno` always returns the number that we need. This conveniently resolves #116987 because no string pattern is checked anymore.

There's no user observable bahavior change so skip the news. (Well it kind of fixes a bug but that's just a coincidence. I'm happy to add a news entry if that's needed)

* Issue: gh-116987

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
#116987 is about the user observable behavior.

I think it is worth to add a NEWS entry.

**Comment by gaogaotiantian** [comment]
Sure, I added the news entry.

File: Misc/NEWS.d/next/Library/2024-03-19-19-42-25.gh-issue-116987.ZVKUH1.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Remove regex pattern check in :func:`inspect.findsource` which makes the function work with class code objects.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2024-03-19-19-42-25.gh-issue-116987.ZVKUH1.rst]: "Remove regex pattern check" is an implementation detail which users do not need to know.

The observable effect is that inspect.findsource now works with class code objects.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by blhsing** [comment]
Looks good to me too @gaogaotiantian. I was looking into fixing the issue myself but was scratching my head over under what circumstances `code.co_firstlineno` can become inaccurate in the first place. Turns out it was a historic thing. Thanks!

File: Lib/inspect.py [hunk 2]
```diff
@@ -1157,15 +1157,8 @@ def findsource(object):
...
-        pat = re.compile(r'^(\s*def\s)|(\s*async\s+def\s)|(.*(?<!\w)lambda(:|\s))|^(\s*@)')
-        while lnum > 0:
-            try:
-                line = lines[lnum]
-            except IndexError:
-                raise OSError('lineno is out of bounds')
-            if pat.match(line):
-                break
-            lnum = lnum - 1
+        if lnum > len(lines):
```
Comment (blhsing) [inline Lib/inspect.py]: `lines[lnum]` would be out of bounds if `lnum == len(lines)` too.

Comment (gaogaotiantian) [reply Lib/inspect.py]: You are right, fixed.

**Comment by serhiy-storchaka** [comment]
Could you please also add a test for a class code object?

**Comment by gaogaotiantian** [comment]
Done.

**Comment by gaogaotiantian** [comment]
There's some glitch on the CI.

**Pull request merged**